0000-zk-framework.md

```diff
@@ -0,0 +1,117 @@
+and match between different backends and frontends. In this zero knowledge framework one can use
+rust r1cs library as front end and either bellman or bulletproof as backend. In future versions we
+plan to support more backends.
+
+Current version of the implementation: https://github.com/hyperledger/ursa/pull/192
+
+# Motivation
+[motivation]: #motivation
+
+Finding the appropriate zero knowledge framework (specifically the backend) is a non-trivial design
+choice for application developers. Different backends have different performance numbers depending
+on deployment options which might not be obvious to the application developer. Moreover, application
+developers can have different preferences for different frontends. To make life easy for application
+developers, ideally we should have seemless compatibility between different frontend and backends. In
+the first version of this framework we support rust r1cs as front end, bellman and bullet proof as
+backend.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This project demonstrates R1CS logic with backends to Bellman and Bulletproofs.
+
+Use the following commands to execute each step of the process. The steps are setup (bellman),
+prove, and verify.
+
+Setup
+
+`cargo run -- setup --backend=bellman | tee params.json`
+
+Prove
+
+`echo '"1234"' > witness.hex`
+`cargo run -- prove --backend=bellman --parameters=params.json --witness=witness.hex | tee
+proof.json`
+
+or 
+
+`cargo run -- prove --backend=bulletproofs --witness=witness.hex | tee proof.json`
+
+Verify
+
+`cargo run -- verify --backend=bellman --parameters=params.json --input=proof.json`
+
+or
+
+`cargo run -- verify --backend=bulletproofs --input=proof.json`
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+An R1CS instance can be defined by three matrices A, B and C. They correspond to the NP-complete
+decision problem: whether there is a witness vector w s.t. Aw x Bw = Cw ?
+
+main.rs: this file contains sample code on how to use the framework, get_circuit function defines
+the circuit for which zero knowledge proofs will be generated
+r1csbellman.rs, r1csbulletproof.rs: modules that convert r1cs spec to underlying backends
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+We have chosen r1cs rust library (https://github.com/mir-protocol/r1cs) as frontend, which might
+not be the preferred frontend for all users.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not
+  choosing them?
+- What is the impact of not doing this?
+- For incorporating new protocol implementations what other implementations
+  exist and why were they not selected?
+- For new protocols, what related protocols exist and why do the not satisfy
+  requirements?
```

Comment:
these are questions that should be answered in the alternatives section. The snippet below is okay, but doesn't go into the detail that answering these questions would.

While I would prefer to have actual answers to these questions, if the text below is intended to be sufficient, then the questions themselves should be removed.